File: trunk/Lib/DotNet/aaaSoft.Net/aaaSoft.Net/Xiep/Helpers/XiepIoHelper.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Net.Sockets;
using aaaSoft.Helpers;
using System.Diagnostics;
using aaaSoft.Net.Xiep.Packages;
using System.IO;

namespace aaaSoft.Net.Xiep.Helpers
{
    /// <summary>
    /// XIEP协议的输入输出辅助类
    /// </summary>
    public class XiepIoHelper
    {
        //传输编码
        public static String transferEncoding = "utf-8";
        //最大接收包大小
        public static int maxReceivePackageSize = 10 * 1024 * 1024;

        //得到Encoding对象
        private static Encoding getEncoding()
        {
            Encoding encoding = new UTF8Encoding(false);
            try
            {
                encoding = Encoding.GetEncoding(transferEncoding);
            }
            catch
            {
            }
            return encoding;
        }

        /// <summary>
        /// 发送XmlTreeNode包
        /// </summary>
        /// <param name="ns"></param>
        /// <param name="node"></param>
        /// <returns></returns>
        public static bool SendPackage(NetworkStream ns, AbstractXiepPackage xiepPackage)
        {
            if (ns == null || xiepPackage == null)
                return false;

            lock (ns)
            {
                String responseNodeString = xiepPackage.ToXml();
                try
                {
                    Byte[] buffer = XiepIoHelper.getEncoding().GetBytes(responseNodeString);
                    //写数据包大小
                    Byte[] packageSizeBuffer = NumberHelper.intToByte(buffer.Length);
                    ns.Write(packageSizeBuffer, 0, packageSizeBuffer.Length);
                    //写数据包内容
                    ns.Write(buffer, 0, buffer.Length);
                    ns.Flush();

                    return true;
                }
                catch
                {
                    return false;
                }
            }
        }

        public static String ReceiveXml(NetworkStream ns)
        {
            try
            {
                int packageSizeNumberBytesCount = 4;
                byte[] packageSizeByteArray = new byte[packageSizeNumberBytesCount];
                //读取包大小
                while (true)
                {
                    int readPackageSizeCount = ns.Read(packageSizeByteArray, 0, packageSizeNumberBytesCount);
                    if (readPackageSizeCount <= 0)
                    {
                        throw new IOException(String.Format("readPackageSizeCount为{0}！", readPackageSizeCount));
                    }
                    if (readPackageSizeCount != packageSizeNumberBytesCount)
                    {
                        Debug.Print("readPackageSizeCount:" + readPackageSizeCount);
                        continue;
                    }
                    break;
                }

                int packageSize = NumberHelper.bytesToInt(packageSizeByteArray);

                if (packageSize <= 0)
                {
                    throw new IOException("包大小不能为负数，已丢弃！", maxReceivePackageSize);
                }
                if (packageSize > maxReceivePackageSize)
                {
                    throw new IOException(String.Format("包大小超过 {0}，已丢弃！", maxReceivePackageSize));
                }

                MemoryStream ms = new MemoryStream(packageSize);
                //读取数据
                IoHelper.CopyStream(ns, ms, Convert.ToInt64(packageSize));
                ms.Position = 0;

                return XiepIoHelper.getEncoding().GetString(ms.ToArray());
            }
            catch
            {
                return null;
            }
        }

        // 接收AbstractXiepPackage包
        public static AbstractXiepPackage ReceivePackage(NetworkStream ns)
        {

            String xml = ReceiveXml(ns);
            if (String.IsNullOrEmpty(xml))
            {
                return null;
            }

            AbstractXiepPackage rtnPackage = null;

            XmlTreeNode treeNode = XmlTreeNode.FromXml(xml);
            String packageName = treeNode.Key;

            if (packageName.Equals("RequestPackage"))
            {
                rtnPackage = new RequestPackage();
            }
            else if (packageName.Equals("ResponsePackage"))
            {
                rtnPackage = new ResponsePackage();
            }
            else if (packageName.Equals("EventPackage"))
            {
                rtnPackage = new EventPackage();
            }
            else
            {
                return null;
            }

            rtnPackage.SetRootPackage(treeNode);
            return rtnPackage;
        }
    }
}

File: Lib/Java/com.scbeta.net/src/com/scbeta/net/xiep/helpers/XiepIoHelper.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package com.scbeta.net.xiep.helpers;

import com.scbeta.helpers.ConsoleHelper;
import com.scbeta.helpers.IoHelper;
import com.scbeta.helpers.NumberHelper;
import com.scbeta.helpers.XmlTreeNode;
import com.scbeta.net.xiep.packages.AbstractXiepPackage;
import com.scbeta.net.xiep.packages.EventPackage;
import com.scbeta.net.xiep.packages.RequestPackage;
import com.scbeta.net.xiep.packages.ResponsePackage;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;

/**
 * XIEP协议网络辅助类
 * @author aaa
 */
public class XiepIoHelper {

    //传输编码
    public static String transferEncoding = "utf-8";
    //最大接收包大小
    public static int maxReceivePackageSize = 10 * 1024 * 1024;

    // 发送XmlTreeNode包
    public static boolean SendPackage(Socket socket, AbstractXiepPackage xiepPackage) {
        synchronized (socket) {
            String responseNodeString = xiepPackage.toXml();
            try {
                OutputStream outputStream = socket.getOutputStream();

                byte[] buffer = responseNodeString.getBytes(transferEncoding);
                byte[] packageSizeByteArray = NumberHelper.intToByte(buffer.length);

                //写数据包大小
                outputStream.write(packageSizeByteArray);
                //写数据包内容
                outputStream.write(buffer);
                outputStream.flush();
                return true;
            } catch (Exception ex) {
                return false;
            }
        }
    }

    // 接收XML字符串
    public static String ReceiveXml(Socket socket) {
        InputStream inputStream = null;
        try {
            inputStream = socket.getInputStream();
        } catch (Exception ex) {
            return null;
        }
        synchronized (inputStream) {
            try {
                int packageSizeNumberBytesCount = 4;

                byte[] packageSizeByteArray = new byte[packageSizeNumberBytesCount];
                //读取包大小
                while (true) {
                    int readPackageSizeCount = inputStream.read(packageSizeByteArray, 0, packageSizeNumberBytesCount);
                    if (readPackageSizeCount < 0) {
                        throw new IOException(String.format("readPackageSizeCount为-1！", maxReceivePackageSize));
                    }
                    if (readPackageSizeCount != packageSizeNumberBytesCount) {
                        ConsoleHelper.println("readPackageSizeCount:" + readPackageSizeCount);
                        continue;
                    }
                    break;
                }

                int packageSize = NumberHelper.bytesToInt(packageSizeByteArray);

                if (packageSize <= 0) {
                    throw new IOException(String.format("包大小不能为负数，已丢弃！", maxReceivePackageSize));
                }
                if (packageSize > maxReceivePackageSize) {
                    throw new IOException(String.format("包大小超过 %s，已丢弃！", maxReceivePackageSize));
                }

                ByteArrayOutputStream outputStream = new ByteArrayOutputStream(packageSize);
                //读取数据
                IoHelper.CopyStream(inputStream, outputStream, packageSize);

                String xmlString = outputStream.toString(transferEncoding);
                //ConsoleHelper.println(String.format("包大小：%s，包内容：%s", packageSize,xmlString));
                return xmlString;
            } catch (IOException ex) {
                return null;
            }
        }
    }

    // 接收AbstractXiepPackage包
    public static AbstractXiepPackage ReceivePackage(Socket socket) {
        String xml = ReceiveXml(socket);
        if (xml == null || xml.isEmpty()) {
            return null;
        }

        AbstractXiepPackage rtnPackage = null;

        XmlTreeNode treeNode = XmlTreeNode.fromXml(xml);
        String packageName = treeNode.getKey();

        if (packageName.equals("RequestPackage")) {
            rtnPackage = new RequestPackage();
        } else if (packageName.equals("ResponsePackage")) {
            rtnPackage = new ResponsePackage();
        } else if (packageName.equals("EventPackage")) {
            rtnPackage = new EventPackage();
        } else {
            return null;
        }

        rtnPackage.setRootPackage(treeNode);
        return rtnPackage;
    }
}
